### PR TITLE
[FF-3332] - Support alternative branch based versioning strategies

### PR DESCRIPTION
**Pull request by credfeto**
# Description

Supports branches using releases like the ones below to cover the strategies used in the games repo:

* release/1.2.3
* release-test/1.2.3
* release-test/game/1.2.3
* release/game/1.2.3

# Related Issue\Feature

FF-3332

# How Has This Been Tested

- [x] All unit tests pass.
- [x] All integration tests pass.
- [x] Manual Testing:

Ran the tool on a couple of local repositories to verify that it comes out with the correct arguments

# Types of changes

- [x] Docs change
- [x] Refactoring
- [ ] Dependency upgrade
- [x] Additional Unit Tests\Integration Tests
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Removed no-longer used code

## Deployment Configuration Changes

* Can no longer pass  Build number as a parameter e.g ``BuildVersion 27``  - now have to use use a named argument ``buildversion --BuildNumber 27`` as there are additional optional parameter.  All builds and meta-runners using this will need modifying.

- [x] Requires deployment configuration changes as specified below and in CHANGELOG.md

# Checklist

- [x] There are no Resharper\static code analysis errors anywhere in the solution.
- [x] I have ONLY run a code clean-up on any files I have modified to make sure they are in the correct format and no others.
- [x] I have added tests to cover my changes.
- [x] I have run the code and quickly verified it all works to my satisfaction.
- [x] All new/modified code has sufficient logging to be able to diagnose what is wrong.
- [x] All new and existing tests passed.
- [x] All new/modified public interfaces/classes have are documented with xmldoc comments.
- [x] Unreleased section of CHANGELOG.md has been updated with details of this PR.

### CODE REVIEW

File: src/FunFair.BuildVersion.Detection/VersionDetector.cs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+            const int maxSuffixLength = 15;
+
+            if (usedSuffix.Length > maxSuffixLength)
+            {
+                usedSuffix = usedSuffix.Substring(startIndex: 0, length: maxSuffixLength);
+            }
+
+            // Ensure that the name doesn't end with a -
+            usedSuffix = usedSuffix.Trim(trimChar: '-');
```
Comment (joshstevens19) [inline src/FunFair.BuildVersion.Detection/VersionDetector.cs:144]: `replacementChar` here?

File: src/FunFair.BuildVersion.Detection/VersionDetector.cs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+            const char replacementChar = '-';
+
+            foreach (char ch in currentBranch.Where(predicate: c => !char.IsLetterOrDigit(c) && c != replacementChar)
+                                             .Distinct())
+            {
+                suffix.Replace(oldChar: ch, newChar: replacementChar);
+            }
+
+            while (suffix.ToString()
+                         .Contains("--"))
```
Comment (joshstevens19) [inline src/FunFair.BuildVersion.Detection/VersionDetector.cs:121]: want to name `--`?

File: src/FunFair.BuildVersion.Detection/VersionDetector.cs [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+            string usedSuffix = this.BuildPreReleaseSuffix(currentBranch: currentBranch);
+
+            this._logger.LogInformation($"Build Pre-Release Suffix: {usedSuffix}");
+
+            Version version = new(major: latest.Version.Major, minor: latest.Version.Minor, latest.Version.Build + 1, revision: buildNumber);
+
+            return new NuGetVersion(version: version, releaseLabel: usedSuffix);
+        }
+
+        private string BuildPreReleaseSuffix(string currentBranch)
```
Comment (joshstevens19) [inline src/FunFair.BuildVersion.Detection/VersionDetector.cs:95]: this method was quite hard to read

**Review by joshstevens19 — CHANGES_REQUESTED** [review]

File: src/FunFair.BuildVersion.Detection/BranchSettings.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    /// </summary>
+    [DebuggerDisplay("Suffix: {ReleaseSuffix} Package: {Package}")]
+    public sealed class BranchSettings : IBranchSettings
+    {
+        /// <summary>
+        ///     Constructor.
+        /// </summary>
+        /// <param name="releaseSuffix">Release suffix.</param>
+        /// <param name="package">Package being released.</param>
+        public BranchSettings(string? releaseSuffix, string? package)
```
[the remark this answers is not in the document]
Comment (credfeto) [reply src/FunFair.BuildVersion.Detection/BranchSettings.cs:17]: Can't as that'll be a DI failure :(

**Review by joshstevens19 — APPROVED** [review]

**Pull request merged**